### PR TITLE
Add nullable option to attributes

### PR DESCRIPTION
**Pull request by wenderjean**
The proposal from PR #67 has originated a refactoring on `coercion` modules (#80) and this new feature.

### CODE REVIEW

File: docs/schema-concept/nullable-attributes.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  name: null
+});
+
+user.attributes; // { name: null }
+user.validate() // { valid: true }
+```
+
+### Accepting nullables (required attributes)
+
+We understand that when an attribute is required it should to be present and actually has a value assigned to it, even an `undefined`, `null` or whatever value what it means that coercion function will not assign a __default__ value even if __nullable__ attribute is __false__.
```
Comment (talyssonoc) [inline docs/schema-concept/nullable-attributes.md]:
```suggestion
We consider that when an attribute is __required___ there should be some value assigned to it even if it's `undefined`, `null` or any other value. It means that coercion will never assign a __default__ value to __required__ attributes even if __nullable__ option is __false__.
```

File: docs/schema-concept/nullable-attributes.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+```javascript
+/*
+ * User.js
+*/
+const User = attributes({
+  name: {
+    type: String,
+    required: true,
+    nullable: false
```
Comment (talyssonoc) [inline docs/schema-concept/nullable-attributes.md]: I think we can add a comment here like `// non-nullable required attribute`

File: docs/schema-concept/nullable-attributes.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+const user = new User({
+  name: null
+});
+
+user.attributes; // { name: null }
+user.validate() // { valid: false }
+```
+
+But notice that you can choose to allow __null__ values on __required__ attributes which will cause the validation to return __true__.
```
Comment (talyssonoc) [inline docs/schema-concept/nullable-attributes.md]:
```suggestion
But notice that you can choose to allow __null__ to be a valid value for __required__ attributes by making it __nullable__, which will cause the validation to return __true__.
```

File: docs/serialization.md [hunk 4]
```diff
@@ -6,6 +6,7 @@ It's possible to obtain a serialized object of a Structure using the method `toJ
 
 - Be aware that `toJSON()` will return an object, not the JSON in form of a string like `JSON.stringify()` does
 - Refer to the [Date#toJSON](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Date/toJSON) specification to see how dates will be serialized by `JSON.stringify`
+- Refer to [Dealing with nullable attributes](schema-concept/nullable-attributes.md) to check how `nullables` are going to be returned on __Serialization__
```
Comment (talyssonoc) [inline docs/serialization.md]: We're having kind of a recursion here 😛, this section tells us to check `nullable-attributes` page, and `nullabled-attributes` page tells us to check it here.

File: docs/schema-concept/nullable-attributes.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  name: null
+});
+
+user.attributes; // { name: null }
+user.validate() // { valid: true }
+```
+
+**Important:**
+
+- Notice that usually an __undefined__ or __null__ value is not returned when you serialize your schema but when you accept a __nullable__ attribute this attribute is going also to be returned in your serialized schema, see [Serialization](serialization.md)
```
Comment (talyssonoc) [inline docs/schema-concept/nullable-attributes.md]:
```suggestion
- Notice that usually an attribute with the value __undefined__ or __null__ is not included when you serialize your structure, but when it is __nullable__ this attribute is going also to be returned in your serialized schema.
```

File: docs/schema-concept/nullable-attributes.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
+# Nullable attributes
+
+You can change the way an attribute is treated when the value `null`is assigned to it by using the `nullable` option with the value `true`, this would affect the way the attribute is coerced, validated and serialized.
```
Comment (kostadriano) [inline docs/schema-concept/nullable-attributes.md]: Missing one space here
```diff
- value `null`is
+ value `null` is 
```
:)

File: docs/schema-concept/nullable-attributes.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+user.attributes; // { name: null }
+user.validate() // { valid: true }
+```
+
+**Important:**
+
+- Notice that by not using the `nullable` option the __default__ value for __String__ is an empty string, which means that you need to accept empty strings to make your schema valid.
+
+- Notice that usually an attribute with the value __undefined__ or __null__ is not included when you serialize your structure, but when it is __nullable__ this attribute is going to be returned in your serialized schema.
```
Comment (talyssonoc) [inline docs/schema-concept/nullable-attributes.md]: What do you think?

```suggestion
- Notice that usually an attribute with the value __null__ is not included when you serialize your structure, but when it is __nullable__ this attribute is going to be returned in your serialized structure even if its value is __null__.
```

File: docs/schema-concept/nullable-attributes.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
+# Nullable attributes
+
+You can change the way an attribute is treated when the value `null` is assigned to it by using the `nullable` option with the value `true`, this would affect the way the attribute is coerced, validated and serialized.
+
+If you do not set the `nullable` option for an attribute it will default to `false` and make your attribute __not nullable__.
+
```
Comment (talyssonoc) [inline docs/schema-concept/nullable-attributes.md]: What do you think about adding a text here:

```suggestion
Here's the behavior without the `nullable` option:
```

File: test/unit/coercion/boolean.spec.js [hunk 9]
```diff
@@ -19,6 +23,14 @@ describe('type coercion', () => {
       expect(user.isAdmin).to.be.undefined;
     });
 
+    it('does not coerces null when nullable', () => {
```
Comment (talyssonoc) [inline test/unit/coercion/boolean.spec.js]:
```suggestion
    it('does not coerce null when nullable', () => {
```

File: test/unit/coercion/date.spec.js [hunk 10]
```diff
@@ -28,6 +32,14 @@ describe('type coercion', () => {
       expect(user.birth).to.be.undefined;
     });
 
+    it('does not coerces null when nullable', () => {
```
Comment (talyssonoc) [inline test/unit/coercion/date.spec.js]:
```suggestion
    it('does not coerce null when nullable', () => {
```

File: test/unit/coercion/string.spec.js [hunk 11]
```diff
@@ -7,68 +7,136 @@ describe('type coercion', () => {
...
-      expect(user.name).to.not.equal('Some name');
-      expect(user.name).to.not.equal(new String('Some name'));
-      expect(user.name).to.eql(name);
-    });
+        expect(user.name).to.not.equal('Some name');
+        expect(user.name).to.not.equal(new String('Some name'));
+        expect(user.name).to.eql(name);
+      });
+
+      it('does not coerces undefined', () => {
```
Comment (talyssonoc) [inline test/unit/coercion/string.spec.js]:
```suggestion
      it('does not coerce undefined', () => {
```

**Pull request merged**